woocommerce/google-listings-and-ads
# API Pull - Verify a nonce when setting Google WPCOM auth status

### Changes proposed in this Pull Request:



This is a follow up from a PR comment: https://github.com/woocommerce/google-listings-and-ads/pull/2311#discussion_r1540827308.

The auth flow:

1. Merchant clicks enable new product sync
2. GL&A calls a new Google API to get the nonce, and add it in the `state` query parameter of the auth URL. Also save the nonce as an option in the DB.
3. GL&A redirects the merchant to the auth URL.
4. Merchant approves the access, then WPCOM will call Google’s callback URL (redirect URI as in the WPCOM OAuth document)
5. Google gets the nonce from the `state` query parameter, verify it, and get merchant’s WPCOM access token. Note that the “return URL” is also included in the `state` parameter.
6. Google redirects back to the merchant site, with query parameter `google_wpcom_app_status` and `nonce`.
7. GL&A verify the nonce parameter with the option in the DB, and save the `google_wpcom_app_status` status in the DB as well.

This PR implements the last step of the above flow, which it compares the nonce query parameter with the nonce option in the DB, and finally store the `google_wpcom_app_status` status in the DB.

### Detailed test instructions:


**Prerequisite**
1. Make sure the notification is enabled by setting `add_filter( 'woocommerce_gla_notifications_enabled', '__return_true' )`.
2. Make sure the options `gla_wpcom_rest_api_status` and `gla_google_wpcom_auth_nonce` are not in the DB.

**Test Flow**
1. Complete the onboarding and go to the settings page.
2. Click `Enable` new product sync from the top-left banner.
3. It should redirect you to the WPCOM app granting access page.
4. Check the DB, the option `gla_google_wpcom_auth_nonce` should be set to `nonce-123`.
    - Note that since we are still waiting for the new Google API, the nonce is hardcoded as `nonce-123` at the moment.
5. Complete the auth flow, and go back to settings page with query parameters `google_wpcom_app_status=approved` and `nonce=nonce-123`, e.g.:
    - `/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fsettings&google_wpcom_app_status=approved&nonce=nonce-123`
6. The GMC card on the settings page should show `Google has been granted access to fetch your product data.`.
7. Check the DB, the option `gla_wpcom_rest_api_status` should be set as `approved`.
8. Check the DB, the option `gla_google_wpcom_auth_nonce` should be removed.
9. Redo the step 2 to 5, and use a wrong nonce query param other than `nonce-123` in the step 5.
10. Make sure the status is not being set, and you can see an error message in the browser's dev console
11. Redo the step 2 to 5, and use the correct nonce and another `google_wpcom_app_status` such as `disapproved` and `error`.
12. Make sure the status is set correctly in the DB and the UI is showing `There was an issue granting access to Google for fetching your products.`

### Additional details:


### Changelog entry

> Tweak - API Pull: Verify a nonce when setting Google WPCOM auth status
